### DISCUSSION TITLE
Usage of BGP over different LANs

### DISCUSSION THREAD

**Discussion by dberardo-com** [techno-tim/k3s-ansible#82] [Questions]
ARP works at layer2, so it would work fine in bare-metal setups running underneath the same router. 

Can MetalLB or Kube-vip be used to achieve VIP assignment over a WAN network? Is this perhaps BGP's job?

Info: i have setup a demo cluster using VPS hosted on different data centers. Those VPS are within an IPsec VPN listening on a "separate" virtual interface. Would it be possible to place a router within the VPN and assign this router as a default gateway to resolve the VIP address within the nodes, so as to achieve a BGP based HA setup?

or what are better solutions out there?

**Comment by dberardo-com** [comment]
here perhaps a similar idea from another source: https://www.arubanetworks.com/techdocs/central/latest/content/sd-branch/cfg/routing/bgp-ipsec-tunnel.htm

**Comment by modzilla99** [comment]
Yes that would be entirely possible. I've been using MetalLB with BGP in the past. It is a rabbit hole though 😂 !  For the VPN you can use WireGuard as well. 


```
             BGP + ECMP over VPN
     +----------------+----------------+
     |                |                |
  vrouter1         vrouter2         vrouter3
     |                |                |
  net1-dc1         net2-dc2         net3-dc3
(10.1.0.0/24)    (10.2.0.0/24)    (10.3.0.0/24)
     |                |                |
  master1          master2          master3
(10.1.0.10)      (10.2.0.10)      (10.3.0.10)
     |                |                |
  worker1          worker2          worker3
(10.1.0.11)      (10.2.0.11)      (10.3.0.11)
```

**Reply by dberardo-com** [reply]
thanks for your valuable contribution, some questions here:

1. how did you draw the amazing map above ? that looks like a tedious job to do on a notepad :D
2. do the master nodes in your map have to "see each other" directly for the cluster to operate? i guess so, right? and how about the workers? is it enought for masters to know about their own workers or do all nodes have to be able to communicate ?
3. what is the net-dc ? is it a software component or an hardware one ? and what is its role here ?
4. so if i understand your map, then MetlaLB is placed at the same level of the master nodes, and broadcasts the VIP allocation to all vrouters at the same time. Is this correct?  do you feel like sharing a sample configuration object (CRD, etc.) of your metalLB setup?

i feel one step closer to the solution, and i am glad to see your message here, since my research had actually lead me to something much broder like: https://www.k8gb.io/

--- 

UPDATE

i understand that if we are talking about Public IPs instead of private IPs as VIP, then we still need a GSLB, or can BGP in such configuration be used also to advertise the same public address on different machines ? Of course, one has to "own" a pool of IP addresses, etc.

**Reply by modzilla99** [reply]
1. I actually did... :D
2. They all need L3 connectivity to each other. L2 is not required, tho.
3. That was just meant to show off the subnet used. It's nothing really.
4. Can't right now unfortunately as that's not my setup. But MetalLB by default advertises the Load balancer Service IP from all nodes running the speaker, resulting in ECMP Routes. No special sauce needed.
5.  (The update one) BGP totally does that. At work we host services in two different locations, while advertising our IP space in both of them to different peers.

**Reply by dberardo-com** [reply]
i see, i am still trying to get my head around your solution, hopefully your answers will guide me further.

as for point number 5: are the peers within the same LAN or do they reach the cluster via public internet, and if so, then i assume you have a sort of floating IP that you have the capability to assign to any one of the 2 locations, right?

i am trying to setup an HA solution also between different sites, but on VPSs hosted by different cloud providers and somehow i am not able to acquire a floatingIP that could be used across those different providers ...

**Reply by modzilla99** [reply]
Yeah tbh it's pretty complex and a lot to understand and learn.

I'm not sure if I can follow your questions regarding num5. You might want to read up on Anycast IPs for that. But basically you have to 2 seperate servers in different colos, but both will advertise the same IP address let's say 8.8.8.8 (Google does exactly that ;) ) But if you try to access that IP from your home it will be routed to the closets location there is, because BGP knows the hops (other routers) in between and prefers the one with the least.

But if you have two servers side by side (either same net or same hops in between) advertising the same IP, your router will install ECMP routes. That stands for equal cost multipath and will layer 4 loadbalance the traffic to those servers.

But to do BGP with your cloud provider you'll need to be part of RIPE for example and own an IP subnet. That's expensive and not feasible for v4. But if you have one VULTR does allow you to peer for example

**Reply by dberardo-com** [reply]
> Yeah tbh it's pretty complex and a lot to understand and learn.

i do see that :D that's why i was trying to find some kind of lower-tech solutions like acting on a DNS level instead (https://www.k8gb.io/) to achieve HA.

> But to do BGP with your cloud provider you'll need to be part of RIPE for example and own an IP subnet. That's expensive and not feasible for v4.

exactly what i was fearing, it seemed to me also that in order to achieve an allocation of the same public IP on a different location one would need some sort of "higher privileges". thats why i guess the only viable options i see here are:

* automated "failover" via the DNS (example k8gb -- i hope they support this use case)
* acquiring an HA-ready external load balancer

> But if you have one VULTR does allow you to peer for example

i see that VULTR also offers load balancing services, but like all cloud providers i know, they only let you load balance backend services that also come from their cloud, so you can't bring in your own data centers or servers from other providers for example (https://www.vultr.com/docs/vultr-load-balancers/)

Would you agree that the 2 options mentioned above are the only viable ones? i understand that the solution you modeled in your schema above is only a viable one within a private network (distributed LAN), without RIPE membership at least. Or am i wrong ?

**Reply by modzilla99** [reply]
Yeah you're right, DNS is the best option for cheap and easy failover. It's takes a long time due to the many layers of caching involved but it's the way to go, if you want/need multiple cloud providers or multiple regions where cloudloadbalancers don't work.

And yes again, that schema I did will only work for internal services if you don't own a subnet. Might be worth a read anyways: https://blog.kroy.io/2020/12/07/vyos-from-scratch-bgp-announcements/

But if you really wanna do it the rules for v6 are more relaxed you'll only need a member that will "sponsor" you. Haven't done it myself (yet) but seems interesting.

**Reply by dberardo-com** [reply]
thanks for confirming my intuitions, i will definitely have a look at the article you posted and see what i can draw out of it.

> But if you really wanna do it the rules for v6 are more relaxed you'll only need a member that will "sponsor" you. Haven't done it myself (yet) but seems interesting.

sounds quite interesting, what do you mean with "sponsor" ?

**Reply by modzilla99** [reply]
Sponsor is the phrase they use for a person/company that will handle all the RIPE related stuff like registering an ASN or reserving an IP block.
